### DISCUSSION TITLE
ICACHE_FLASH_ATTR

### DISCUSSION THREAD

**Discussion by benoitLecuppe** [micropython/micropython#10617] [ESP8266] [answered]
Bonjour,

I'm working on esppwm.c.  Most functions have the attribute ICACHE_FLASH_ATTR  . 
This attribute indicates to the linker to put the code in .irom0.text section.
if I undefine ICACHE_FLASH_ATTR, I have an error  "build-GENERIC/firmware.elf section `.text' will not fit in region `iram1_0_seg'".
My question is : how the other modules, like hspi.c for exemple , are putting the code in .irom.text section without using ICACHE_FLASH_ATTR   ?

Merci :-)
Benoit

**Comment by jimmo** [comment, marked as answer]
@benoitLecuppe

This is a bit subtle... `esp8266_common.ld` will place anything by default into `iram1_0_seg`. However there are a list of specific object files above that are explicitly placed in `irom0_0_seg`. (This includes `hspi.o` but does not include `esppwm.o`).

So `ICACHE_FLASH_ATTR` is used explicitly in `esppwm.c` to place certain functions into `.irom0.text` (i.e. `irom0_0_seg`) because by default everything in that file is going into `iram1_0_seg`. When you undefine that macro, everything in that file now goes to `iram1_0_seg`, overflowing it.

**Comment by benoitLecuppe** [comment]
Merci,

I understand. I  put esppwm.o in esp8266_common.ld and It's work. 
Have you an idea of the reason why esppwm.o is not in esp8266_common.ld ?

**Reply by jimmo** [reply]
> Have you an idea of the reason why esppwm.o is not in esp8266_common.ld ?

@benoitLecuppe Not sure, sorry. I think it would probably be clearer if it was and then explicitly marked the functions that should be in `iram`.
